### PR TITLE
[Feature] Add Search-Transformer.

### PR DESCRIPTION
**Pull request by Yshuo-Li**
For TTSR

### CODE REVIEW

File: mmedit/models/transformers/search_transformer.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            outputs (Tensor): The result tensor. (N, C*k*k, H*W)
+        """
+        views = [inputs.size(0)] + [
+            1 if i != dim else -1 for i in range(1, len(inputs.size()))
+        ]
+        expanse = list(inputs.size())
+        expanse[0] = -1
+        expanse[dim] = -1
+        index = index.view(views).expand(expanse)
+        outputs = torch.gather(inputs, dim, index)
```
Comment (innerlee) [inline mmedit/models/transformers/search_transformer.py:34]:
```python
outputs = torch.gather(inputs, dim, index[:, None, :])
```

File: mmedit/models/transformers/search_transformer.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+        Args:
+            inputs (Tensor): The source tensor. (N, C*k*k, H*W)
+            dim (int): The axis along which to index.
+            index (Tensor): The indices of elements to gather. (N, H*W)
+
+        results:
+            outputs (Tensor): The result tensor. (N, C*k*k, H*W)
+        """
+        views = [inputs.size(0)] + [
```
Comment (innerlee) [inline mmedit/models/transformers/search_transformer.py:27]:
```python
views = [-1 if i in (0, dim) else 1 for i in range(inputs.ndim())]

File: mmedit/models/transformers/search_transformer.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            1 if i != dim else -1 for i in range(1, len(inputs.size()))
+        ]
+        expanse = list(inputs.size())
```
Comment (innerlee) [inline mmedit/models/transformers/search_transformer.py:30]:
```python
expansion = [-1 if i in (0, dim) else d for i, d in enumerate(inputs.size())]
```

File: mmedit/models/transformers/search_transformer.py (hunk 1 above)
Comment (Yshuo-Li) [reply mmedit/models/transformers/search_transformer.py:34]: Different results.

**Review by innerlee — APPROVED** [review]

File: mmedit/models/transformers/search_transformer.py (hunk 3 above)
Comment (Yshuo-Li) [reply mmedit/models/transformers/search_transformer.py]: OK, thanks

File: mmedit/models/transformers/search_transformer.py (hunk 2 above)
Comment (Yshuo-Li) [reply mmedit/models/transformers/search_transformer.py:27]: Will cause error.

File: mmedit/models/transformers/search_transformer.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+        key_t = F.normalize(key_t, dim=2)  # [N, Hr*Wr, C*k*k]
+        query = F.normalize(query, dim=1)  # [N, C*k*k, H*W]
+
+        # Relevance embedding
+        rel_embedding = torch.bmm(key_t, query)  # [N, Hr*Wr, H*W]
+        max_val, max_index = torch.max(rel_embedding, dim=1)  # [N, H*W]
+
+        # hard-attention
+        t_level3_unfold = self.gather(value_level3, 2, max_index)
```
Comment (innerlee) [inline mmedit/models/transformers/search_transformer.py:99]: What's the shape of value_level3?

File: mmedit/models/transformers/search_transformer.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            inputs (Tensor): The source tensor. (N, C*k*k, H*W)
+            dim (int): The axis along which to index.
+            index (Tensor): The indices of elements to gather. (N, H*W)
+
+        results:
+            outputs (Tensor): The result tensor. (N, C*k*k, H*W)
+        """
+
+        views = [inputs.size(0)] + [
+            1 if i != dim else -1 for i in range(1, len(inputs.size()))
```
Comment (innerlee) [inline mmedit/models/transformers/search_transformer.py:29]: len(inputs.size()) -> inputs.ndim

File: mmedit/models/transformers/search_transformer.py (hunk 4 above)
Comment (Yshuo-Li) [reply mmedit/models/transformers/search_transformer.py:99]: [N, C*k*k, H*W]

Comment (innerlee) [reply mmedit/models/transformers/search_transformer.py:98]: Okay so Hr == H and Wr == W?

Comment (Yshuo-Li) [reply mmedit/models/transformers/search_transformer.py:98]: Yes

File: mmedit/models/transformers/search_transformer.py [hunk 6]
```diff
@@ -35,31 +35,31 @@ def gather(self, inputs, dim, index):
...
         return outputs
 
-    def forward(self, lr_pad_level3, ref_pad_level3, ref_level1, ref_level2,
+    def forward(self, lq_up_level3, ref_downup_level3, ref_level1, ref_level2,
                 ref_level3):
         """Texture transformer
 
-        Q = LTE(lr_pad)
-        K = LTE(ref_pad)
+        Q = LTE(lq_pad)
```
Comment (innerlee) [inline mmedit/models/transformers/search_transformer.py:42]: pad

**Review by innerlee — APPROVED** [review]
